homeassistant/components/influxdb/__init__.py

```diff
@@ -478,42 +481,95 @@ def close_v1():
-        _LOGGER.error(RETRY_MESSAGE, exc)
-        event_helper.call_later(
-            hass, RETRY_INTERVAL, lambda _: _retry_setup(hass, config)
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=conf,
+            )
         )
-        return True
 
-    event_to_json = _generate_event_to_json(conf)
-    max_tries = conf.get(CONF_RETRY_COUNT)
-    instance = hass.data[DOMAIN] = InfluxThread(hass, influx, event_to_json, max_tries)
-    instance.start()
+    return True
 
-    def shutdown(event):
-        """Shut down the thread."""
-        instance.queue.put(None)
-        instance.join()
-        influx.close()
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
+async def async_setup_entry(hass: HomeAssistant, entry: InfluxDBConfigEntry) -> bool:
+    """Set up InfluxDB from a config entry."""
+    data = entry.data
+
+    hass_config = await conf_util.async_hass_config_yaml(hass)
+
+    influx_yaml = CONFIG_SCHEMA(hass_config).get(DOMAIN, {})
+    default_yaml: dict[str, Any] = {
+        "entity_globs": [],
+        "entities": [],
+        "domains": [],
+    }
+
+    options = {
+        CONF_RETRY_COUNT: influx_yaml.get(CONF_RETRY_COUNT, 0),
+        CONF_PRECISION: influx_yaml.get(CONF_PRECISION),
+        CONF_MEASUREMENT_ATTR: influx_yaml.get(
+            CONF_MEASUREMENT_ATTR, DEFAULT_MEASUREMENT_ATTR
+        ),
+        CONF_DEFAULT_MEASUREMENT: influx_yaml.get(CONF_DEFAULT_MEASUREMENT),
+        CONF_OVERRIDE_MEASUREMENT: influx_yaml.get(CONF_OVERRIDE_MEASUREMENT),
+        CONF_INCLUDE: influx_yaml.get(CONF_INCLUDE, default_yaml),
+        CONF_EXCLUDE: influx_yaml.get(CONF_EXCLUDE, default_yaml),
+        CONF_TAGS: influx_yaml.get(CONF_TAGS, {}),
+        CONF_TAGS_ATTRIBUTES: influx_yaml.get(CONF_TAGS_ATTRIBUTES, []),
+        CONF_IGNORE_ATTRIBUTES: influx_yaml.get(CONF_IGNORE_ATTRIBUTES, []),
+        CONF_COMPONENT_CONFIG: influx_yaml.get(CONF_COMPONENT_CONFIG, {}),
+        CONF_COMPONENT_CONFIG_DOMAIN: influx_yaml.get(CONF_COMPONENT_CONFIG_DOMAIN, {}),
+        CONF_COMPONENT_CONFIG_GLOB: influx_yaml.get(CONF_COMPONENT_CONFIG_GLOB, {}),
+    }
+
+    config = data | options
+
+    try:
+        influx = await hass.async_add_executor_job(get_influx_connection, config, True)
+    except ConnectionError as err:
+        raise ConfigEntryNotReady(err) from err
+
+    event_to_json = _generate_event_to_json(config)
+    max_tries = config[CONF_RETRY_COUNT]
+    influx_thread = hass.data[DOMAIN] = InfluxThread(
+        hass, entry, influx, event_to_json, max_tries
+    )
+    influx_thread.start()
+
+    entry.runtime_data = influx_thread
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: InfluxDBConfigEntry) -> bool:
+    """Unload a config entry."""
+    influx_thread = entry.runtime_data
+
+    influx_thread.shutdown()
```

Comment:
This is not async safe, wrap it in an executor, and await the executor job from an asyncio.timeout context manager to avoid hanging this task forever.